### PR TITLE
approle - fix metadata for generated secret IDs, re-add `wrap_ttl`

### LINKED ISSUE #683: generate_secret_id doesnt have wrap_ttl input
I have been using hvac=0.10.4 for some time now, and has shifted to 0.10.8.
Saw some deprecation warning wrt approle authentication methods, and wanted to use the new methods.

But for create_role_secret_id method alternative, which is generate_secret_id in approle.py, the new method doesnt accept wrap_ttl value. 

Old Method
   ```
 @utils.deprecated_method(
        to_be_removed_in_version='0.12.0',
        new_method=api.auth_methods.AppRole.generate_secret_id,
    )
    def create_role_secret_id(self, role_name, meta=None, cidr_list=None, token_bound_cidrs=None, wrap_ttl=None, mount_point='approle'):
        """POST /auth/<mount_point>/role/<role name>/secret-id

```
New Method

```
    def generate_secret_id(self, role_name, metadata=None, cidr_list=None, token_bound_cidrs=None, mount_point=DEFAULT_MOUNT_POINT):
        """
        Generates and issues a new Secret ID on a role in the auth method.

        Supported methods:
            POST: /auth/{mount_point}/role/{role_name}/secret-id. Produces: 200 application/json
```

Is there a reason why this functionality has been dropped in the new method? 
Vault accepts wrap_ttl to be given over POST.

### LINKED ISSUE #781: approle generate_secret_id(): metadata are always provided, even if empty
Function generate_secret_id is always include metadata in Vault request, even it is empty.

Code like this:
```
secret_request = client.auth.approle.generate_secret_id(
    role_name=role_name,
)
```

produce the following request to Vault:
```json
{ "request": {
...
    "data": {
      "metadata": null
    },
...
}
```

Usually it has no any side effect, unless you have restriction policy with the only allowed parameters, not included the ```metadata``` itself:
```hcl
path "auth/approle/role/client-role-*" {
  capabilities = ["create", "update"]
  allowed_parameters = {
    "token_policies" = ["client-policy"]
    "token_ttl" = ["30s"]
    "token_max_ttl" = ["1h"]
  }
}
```

Then you always got exception:
```
hvac.exceptions.Forbidden: 1 error occurred:
	* permission denied

, on post https://my.vault.org/v1/auth/approle/role/client-role-1/secret-id
```

The fix is easy - do not provide ```metadata``` parameter, if it empty.

### LINKED ISSUE #950: approle.create_custom_secret_id() metadata parameter doesn't work
Similar to: 
- #782 

The metadata handling for this method is broken, as it doesn't serialize the value to JSON. In theory it would work if a JSON string was sent through, but in that case it would fail parameter validation anyway.

### PR DESCRIPTION
**Pull request by Prividen**
Fixes #781 
Fixes #950
Fixes #683 
Closes #778

### CODE REVIEW

**Review by briantist — APPROVED** [review]
I don't have commit rights or anything but this looks good to me.

@Prividen  use "Fixes #781" in your description so that GitHub properly links them.

**Comment by briantist** [comment]
> I don't have commit rights or anything but this looks good to me.
> 
> @Prividen use "Fixes #781" in your description so that GitHub properly links them.

Update: I'm a maintainer now (funny how that happens) 😅

I've made some updates to the PR:
- rebased
- updated the unit and integration tests to prevent regression
- also addressed `create_custom_secret_id` which suffered from even worse bug (#950)

Thank you for the original issue and the PR @Prividen ! Sorry it took so long to get addressed.

**Review by briantist — APPROVED** [review]

**Review by briantist — APPROVED** [review]

File: hvac/api/auth_methods/approle.py [hunk 1]
```diff
@@ -286,7 +292,7 @@ def generate_secret_id(
             mount_point=mount_point,
             role_name=role_name,
         )
-        return self._adapter.post(url=api_path, json=params)
+        return self._adapter.post(url=api_path, json=params, wrap_ttl=wrap_ttl)
```
Comment (dhuckins) [inline hvac/api/auth_methods/approle.py:295]: should `wrap_ttl` be part of the payload? I don't see a `wrap_ttl` parameter on [requests](https://github.com/python/typeshed/blob/b0dc6a340523638df63e1790741689fda693a297/stubs/requests/requests/sessions.pyi#L215-L233)

Comment (briantist) [reply hvac/api/auth_methods/approle.py:295]: it shouldn't be in the payload in this case. `_adapter` is our adapter rather an adapter from `requests`. The `post` method will call the `request` method:
https://github.com/hvac/hvac/blob/develop/hvac/adapters.py#L112-L123

Our `request` method looks for `wrap_ttl` and if it's supplied, it pops it out of the kwargs and makes it into a header:
https://github.com/hvac/hvac/blob/develop/hvac/adapters.py#L288-L290

Comment (briantist) [reply hvac/api/auth_methods/approle.py:295]: I also added both integration and unit test coverage for `wrap_ttl`, if you want to see it in action

**Review by dhuckins — APPROVED** [review]

**Pull request merged**